hibernate-core/src/main/java/org/hibernate/query/sqm/tree/domain/AbstractSqmJoin.java

```diff
@@ -151,4 +153,17 @@ public <X> SqmEntityJoin<R, X> join(Class<X> targetEntityClass) {
 	public <X> SqmEntityJoin<R, X> join(Class<X> targetEntityClass, SqmJoinType joinType) {
 		return super.join( targetEntityClass, joinType );
 	}
+
+	@Override
+	public boolean equals(Object object) {
+		return object instanceof AbstractSqmJoin<?, ?> that
+			&& super.equals( object )
+			&& this.joinType == that.joinType; // unnecessary, but harmless
+//			&& Objects.equals( onClausePredicate, that.onClausePredicate ); // including this causes problems
```

Comment:
I guess this causes a `StackOverflowError` because the predicate usually contains a reference to the entity join again? We'd still need to check equality though, but maybe we need to use `toHqlString()` instead?
Such reentrancy problems are IMO usually best tackled with a custom "protocol". I.e. in `appendHqlString`, I added a `SqmRenderContext` parameter to be able to check substitutability. Maybe the `equals` implementations of the SQM nodes should rather delegate to such a custom equals method with a context parameter. You could then call that method here to avoid doing a deep equals check for `From` nodes within the predicate.

Reply:
Ooooh, wait wait wait. It's completely fine actually.

Casting my mind back, I remember now. This implementation of `equals()` is only used when comparing path expressions, I believe. Not when comparing from clauses. It's actually `SqmFromClause.equalsJoins()` that has this responsibility.

I will add a comment to that effect.